### DISCUSSION TITLE
ByteBuddyAgent.installExternal hangs forever

### DISCUSSION THREAD

**Discussion by ferienc** [raphw/byte-buddy#1629] [Q&A]
Hi everybody,

  long story. Im trying to extend precision of some ORM entitites from `java.lang.Integer` to `java.lang.Long`. But stere are still hidden "backdoors" like `Long.intValue()` which could cause precision loose. So I said to my self be smart use bytebuddy to instrument code. So far so good this peace of code work charmly and check whether calling `Long.intValue()`  dont cause loosing precision.
     
```  
   private static void instrumentImpl() throws IOException, UnmodifiableClassException {   
        Instrumentation instrumentation = ByteBuddyAgent.install();
        File temp = injectBootstrapClasses(instrumentation, LongInterceptor.class);

        new AgentBuilder.Default()
                .disableClassFormatChanges()
                .with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
                .with(AgentBuilder.RedefinitionStrategy.DiscoveryStrategy.Reiterating.INSTANCE)
                .with(AgentBuilder.InitializationStrategy.NoOp.INSTANCE)
                .with(AgentBuilder.TypeStrategy.Default.REBASE)
                .ignore(ElementMatchers.nameStartsWith("net.bytebuddy."))
                //.enableBootstrapInjection(instrumentation, temp)
                .ignore(ElementMatchers.none())
                .type(ElementMatchers.named("java.lang.Long"))
                .transform((builder, t, c, m, p) -> builder.visit(Advice.to(interceptor).on(ElementMatchers.named("intValue"))))
                .installOn(instrumentation);
                //        .installOnByteBuddyAgent();

        instrumentation.retransformClasses(Long.class);
   }

    public static class LongInterceptor {

        @Advice.OnMethodEnter(inline = false)
        public static void enter(@Advice.This Long thisObject) {
            if (thisObject.longValue() != (int) thisObject.longValue()) {
                throw new IllegalStateException("Loosing presision for long value " + thisObject.longValue());
            }
        }
    }

  private static File injectBootstrapClasses(Instrumentation instrumentation, Class<?>... classes) throws IOException {
        File temp = Files.createTempDirectory("tmp").toFile();
        temp.deleteOnExit();

        Map<TypeDescription.ForLoadedType, byte[]> types = Arrays.stream(classes).collect(
                Collectors.toMap(TypeDescription.ForLoadedType::new, ClassFileLocator.ForClassLoader::read));
        ClassInjector.UsingInstrumentation.of(temp, ClassInjector.UsingInstrumentation.Target.BOOTSTRAP, instrumentation).inject(types);

        return temp;
    }
```
   Its kind of magic, but somehow it **works, but only when run as standalone Java app**. Problem come when I tried to bundle it into ear. application in WildFly application server. Seems that problem is in "class loading" which fails in calling `
net.bytebuddy.agent.ByteBuddyAgent#doGetInstrumentation` more precisely in `Class<?> installer = Class.forName(Installer.class.getName(), true, ClassLoader.getSystemClassLoader());`  which silently fails with `java.lang.ClassNotFoundException`, problem is probably in that I cannot find way how to add bytebuddy-agent into system class loader.  Ok not good, but is still should work because bytebuddy seems to use some fallback mechanism which I call "external attachment". But this approach **hangs forever**, here is stack trace 

```
MSC service thread 1-3" #35 prio=5 os_prio=0 cpu=1500.00ms elapsed=170.37s tid=0x0000016d7ecaf810 nid=0x4da4 runnable  [0x00000047095fd000]
   java.lang.Thread.State: RUNNABLE
        at java.lang.ProcessImpl.waitForInterruptibly(java.base@17.0.10/Native Method)
        at java.lang.ProcessImpl.waitFor(java.base@17.0.10/ProcessImpl.java:566)
        at net.bytebuddy.agent.ByteBuddyAgent.installExternal(ByteBuddyAgent.java:705)
        at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:636)
        at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:616)
        - locked <0x0000000716bb0890> (a java.lang.Class for net.bytebuddy.agent.ByteBuddyAgent)
        at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:568)
        at net.bytebuddy.agent.ByteBuddyAgent.install(ByteBuddyAgent.java:545)
        at cz.tipsport.common.util.BootstrapAgent.instrumentImpl(BootstrapAgent.java:50)
```

Pls Any idea whats wrong?? Im digging but not finding any solution.

**Comment by raphw** [comment]
Can you set this property? https://github.com/raphw/byte-buddy/blob/master/byte-buddy-agent/src/main/java/net/bytebuddy/agent/Attacher.java#L33

If something fails in the external process, the stack would be written to the specified file.

**Reply by ferienc** [reply]
Hi @raphw 

  first of all I deeply appreciate how fast you react and how your answer hold the topic. Now to the problem. Thanks for hint, but all of this acts little bit unpredictable. Now it goes another line, not using "external process" to attach JVM but "internal". Im not aware changing anything just wait for two days ;]  Anyway currently attaching works, so pls consider it as done. 
  
![image](https://github.com/raphw/byte-buddy/assets/50740300/5d28f9ac-1679-463f-a9cb-305995b754f4)

**Reply by ferienc** [reply]
look like problem was in absence of `-Djdk.attach.allowAttachSelf=true`, but still, without it JVM should be accessible through external process. Extenal Process started and I can see Java process in task manager but really look like running forever. I added `-Dnet.bytebuddy.agent.attacher.dump`  but I cant see any dump

**Reply by raphw** [reply]
You need to specify a location; -Dnet.bytebuddy.agent.attacher.dump=/my/folder
